core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -285,9 +285,10 @@ private[spark] object Utils extends Logging {
    */
   def createDirectory(dir: File): Boolean = {
     try {
-      // This sporadically fails - not sure why ... !dir.exists() && !dir.mkdirs()
-      // So attempting to create and then check if directory was created or not.
-      dir.mkdirs()
+      // SPARK-35907
+      // This could throw more meaningful exception information if directory creation failed.
+      // To be on the safe side, try to create and then check if directory was created or not.
+      Files.createDirectories(dir.toPath)
       if ( !dir.exists() || !dir.isDirectory) {
         logError(s"Failed to create directory " + dir)
       }
```

Comment:
@Ngone51 I understand you requested this check to be brought back in your [previous comment](https://github.com/apache/spark/pull/33101#discussion_r661245031), but it doesn't seem right to me. Based on the [Javadoc for createDirectories](https://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#createDirectories(java.nio.file.Path,%20java.nio.file.attribute.FileAttribute...)), the contract is that the directory will be created successfully, or an exception will be thrown. Keeping the `exists`/`isDirectory` checks after is redundant. If we're worried about a JDK bug as [described here](https://github.com/apache/spark/pull/33101#discussion_r661579528), why do we trust the implementation of `File#exists()` and `File#isDirectory()` but not `Files.createDirectories()`? IMO the whole point of switching from `File` to the NIO APIs is that they have more sensible semantics and so we don't need to jump through hoops like this.

Reply:
>  Based on the Javadoc for createDirectories, the contract is that the directory will be created successfully, or an exception will be thrown.

I didn't find such a contract there, but only " Unlike the createDirectory method, an exception is not thrown if the directory could not be created because it already exists."...

> If we're worried about a JDK bug as described here, why do we trust the implementation of File#exists() and File#isDirectory() but not Files.createDirectories()?

It's not about which method we should trust but which way is safer to go.  If there's such a contract as you mentioned, I agree we don't need the check. Otherwise, I think it's safer to follow the original way to avoid suffering from the old issue.

Reply:
The word "contract" is not used, but I think that is rarely the case in method descriptions. For example none of the methods on `Dataset` explicitly state contracts or guarantees, but we are expected to assume that the method will do what it says it does, and indicate an error condition using an exception otherwise. My impression here is that the weird semantics of `File#mkdirs()` are translating into FUD about the potential shortcomings of `Files.createDirectories()` and that, if the method were taken without that historical context, we wouldn't be worrying about it.

This all being said, it's only three lines of code extra, so of course it is no big detriment to be on the safe side.

Reply:
> The word "contract" is not used, but I think that is rarely the case in method descriptions. For example none of the methods on `Dataset` explicitly state contracts or guarantees, but we are expected to assume that the method will do what it says it does, and indicate an error condition using an exception otherwise. My impression here is that the weird semantics of `File#mkdirs()` are translating into FUD about the potential shortcomings of `Files.createDirectories()` and that, if the method were taken without that historical context, we wouldn't be worrying about it.
> 
> This all being said, it's only three lines of code extra, so of course it is no big detriment to be on the safe side.

I agree with you in terms of safety and minimizing modification points.Do you have any questions with my current code?

Reply:
@Shockang Could you leave a comment above the check to give the historical context, e.g.,

"The check was required by File#mkdirs() because it could sporadically fail silently. After switching to  Files.createDirectories(), ideally, there should no longer be silent fails. But the check is kept for the safety concern. We can remove the check when we're sure that Files.createDirectories() would never fail silently."

Reply:
> Do you have any questions with my current code?

No, just wanted to explain my viewpoint. Current code is fine. I agree with @Ngone51 that a comment is helpful to explain the context.

Reply:
> @Shockang Could you leave a comment above the check to give the historical context, e.g.,
> 
> "The check was required by File#mkdirs() because it could sporadically fail silently. After switching to Files.createDirectories(), ideally, there should no longer be silent fails. But the check is kept for the safety concern. We can remove the check when we're sure that Files.createDirectories() would never fail silently."

This comment looks perfect. Thank you very much for your comments.